### ISSUE TITLE
migrate hard-wired config paths re Poetry build

### ISSUE THREAD

**Issue by phillxnet** [rockstor/rockstor-core#2540]
Our new Poetry build system (testing branch only) uses the more 'standard' .venv/bin dir to host our local Python env. Previously we used bin within our install directory via a buildout build setup(#2434). Ensure, via initrock (rockstor-pre.service), that we migrate all existing configuration files over to this new path. This could take the form of a new 'establish' element that looks to all relevant in-play config and in-place edits/migrates to ensure the legacy ./bin path is transitioned to the new .venv/bin path. We don't want to simply rely on a fs redirect (symlink) here as it is important to maintain transparency in our error reports. A symlink would obfuscate info in this case. 

We have already had a number of code changes to fix new config elements, but have yet to deal with our prior Stable to proposed Stable (4.1.0-0 to 5.6.0-0 or there abouts) upgrades: a situation that is gaining ergency as we are now in late RC status within the testing branch.

Adding this more specific issue to replace the major part of #2535. Thanks to @stefanogreg  & @FroggyFlox & @Hooverdan96 for exploring this issue here on GitHub and on our forum.

Known remaining configurations affected by this change:

- [x] Samba - we hard-wire within smb.conf a script with the old/non-existent path. I.e. ```root preexec = "/opt/rockstor/bin/mnt-share MyShare"``` to ```root preexec = "/opt/rockstor/.venv/bin/mnt-share MyShare"```
- [x] We have cron elements configured,via Web-UI schedules, that similarly have hard-wired legacy paths to our prior env.

**Comment by phillxnet** [comment]
@FroggyFlox & @Hooverdan96
Please feel free to edit / suggest our above list of known legacy hard-wired config paths.

**Comment by Hooverdan96** [comment]
there is also a path specified in `/opt/rockstor/etc/init.d/gunicorn`

```
PATH=/bin:/usr/bin:/sbin:/usr/sbin
WORKERS=2
PIDFILE=/run/gunicorn.pid
DAEMON=/opt/rockstor/bin/gunicorn
PYPATH=/opt/rockstor/src/rockstor
BIND=127.0.0.1:8000
USER=rocky
LOGFILE=/opt/rockstor/var/log/gunicorn.log
```
I assume, that needs to be addressed as well?


More cosmetic, another search yielded a few more instances where we probably want the wording cleaned up to indicate the `.venv` directory being in between `rockstor` and `bin`

https://github.com/rockstor/rockstor-core/blob/cde01f25f766225bace8e64ed9492a51c636a596/src/rockstor/system/tests/test_directory_services.py#L24-L26

or 
https://github.com/rockstor/rockstor-core/blob/cde01f25f766225bace8e64ed9492a51c636a596/src/rockstor/system/tests/test_system_network.py#L24-L26

I guess in most of these tests we would want to update that comment to include the poetry change? That's something I can certainly quickly do since that's more busywork than anything else. 

If you want to settle on the wording, I can update all the relevant test files, one less thing that would be forgotten as it doesn't _add_ to the functionality per se. Let me know.

**Comment by phillxnet** [comment]
@Hooverdan96 Thanks for the attention on this issue.
Re:

> so, not sure whether this functionally really matters, but should the path be adjusted above?

I think that file is actually legacy now, given we launch & config via supervisord as per:

> Here I can see that gunicorn is launched with the correct path:

So likely we need to wipe that file entirely actually. This was likely a missed clean-up step by me in: #2434 as it had been legacy even before that I believe.

**Comment by phillxnet** [comment]
@Hooverdan96 
Re:

> I guess in most of these tests we would want to update that comment to include the poetry change?

Agreed. I've been doing them as I go along actually, but have a tree wide change to all tests would get that bit sorted and be far less frustrating for folks trying to follow these old path instructions when working on the tests.

See: https://github.com/rockstor/rockstor-core/pull/2500/files
Where I've popped in the half-way-house of:
https://github.com/rockstor/rockstor-core/blob/cde01f25f766225bace8e64ed9492a51c636a596/src/rockstor/storageadmin/tests/test_pools.py#L43-L51

So we also need to update the instructions on re-creating the fixtures also.
In that example it looks like I left an old path for both actually. But you see the new form I used that apparently work.
I prefer this more explicit form of calling via "poetry run django-admin test ..."
as folks can see that we are invoking via poetry env and running the standard django wrapper of django-admin to do the 'test' command.

**Comment by phillxnet** [comment]
@Hooverdan96 Here we have yet more inconsistency on my part! But a cleaner set of instructions that also include the modern fixture creation command.

https://github.com/rockstor/rockstor-core/blob/cde01f25f766225bace8e64ed9492a51c636a596/src/rockstor/storageadmin/tests/test_config_backup.py#L40-L48

**Comment by FroggyFlox** [comment]
> Here we have yet more inconsistency on ~my~ [EDIT] our part!

That one is my fault... sorry :(

I also just realized we already have an issue dedicated to the docstrings in our test files: https://github.com/rockstor/rockstor-core/issues/2475
I'll link to here there.

**Comment by phillxnet** [comment]
@FroggyFlox & @Hooverdan96  Thanks for the info links etc back to that dedicated issue re docstrings on tests. As such I think for this issue we address only functional migrations (re binary changed paths) from our 4.1.0-0 last stable so that we can have folks upgrade from that last Stable instance to our coming stable instances. The dev comment instructions in testing are a bit of a stretch for this issue I think. Especially give that just linked docstrings issue.

**Comment by FroggyFlox** [comment]
> I think for this issue we address only functional migrations (re binary changed paths) from our 4.1.0-0 last stable so that we can have folks upgrade from that last Stable instance to our coming stable instances.

Agreed.

**Comment by FroggyFlox** [comment]
Based on the list in the first message (yet to be edited accordingly), and based on the work done in #2476, we should need to ensure proper path to the binaries in:
- `/etc/cron.d/rockstortab`: scheduled tasks
- `/etc/cron.d/replicationtab`: replication service
- `/etc/samba/smb.conf `

The only other places of which I can think that are NOT directly part of Rockstor's core code would be our systemd unit files. I do believe we already ensure these are proper at boot but I'll have to verify that.

**Comment by FroggyFlox** [comment]
With regards to "updating" `smb.conf`. I see 2 options so far:
- checking the file, and replacing any instance of `root preexec = "/opt/rockstor/bin/mnt-helper ..."
- using our existing code in `system.samba` to refresh our samba config. This would essentially:
    - get all entries in the `SambaShare` model
    - write a new temp file with:
        - global samba config section
        - each share to be exported section. This is where the new path will be written
    - copy this file to `/etc/samba/smb.conf`, thereby overwriting any updated config.

The second option does sound interesting as we don't write any specific file manipulation that would need to updated eventually in the future if we change again the path to `mnt-helper` of `smb.conf`, etc... HOWEVER, this involves getting all SambaShares from the database first and I'm not sure how reliable and robust this would be when ran in `initrock.py`. The first option, however rudimentary, would likely be "safer".

Note that both options would then need to check if the Samba service is enabled and restart it accordingly.

Any input?

**Comment by phillxnet** [comment]
@FroggyFlox Re:

> I do believe we already ensure these are proper at boot but I'll have to verify that.

Yes as I understand this we do check in initrock that our conf/ copies of the same are currently represented on the system, via md5 checks. And so each upgrade will re-assert it's local conf/ copies in the greater system/systemd with an appropriate systemd reload there-after if a change was required. See the meta procedure here:

https://github.com/rockstor/rockstor-core/blob/bdc3df2762ceddf355b482ca50f698654e342861/src/rockstor/scripts/initrock.py#L346-L362

And this code has recently had a re-vamp to ease our soon to be dual branch development of both master and testing. And this re-vamp was part of our accomodating newer systemd techniques such as the override we now use for nginx.

**Comment by phillxnet** [comment]
@FroggyFlox & @Hooverdan96 Re:

> To tackle this, I see 2 options so far:

Agreed: and I see the following as less risky also:

> * checking the file, and replacing any instance of `root preexec = "/opt/rockstor/bin/mnt-helper ..."`

Agreed that a whole sale re-write re-instantiation of our config could have uses in the future, but it could also have some as-yet unforseen issues re for example reace conditions or the like with our own services. Remember that we are, ourselves, running under systemd: we have to tread carefully with regard to managing other systemd servcies. We are already fairly heavy handed and I think given we are only wishing to address a path issue we should stick to addressing that path issue alone. I.e. search for existence of old path and if not found leave well alone - else we are for-ever re-writing a config file for not reason potentially. But if an old path exists we must rectify just that issue and in an explicit fashion. Hence the suggestion to add an additional "establish_poetry_paths" type arrangement within initrock. We might even go so far as to call a shell sed to do this if it is found to be required. Not sure if sed can tell us if it has replaced via rc code or not but that may help. We are just after the simplest form of a path migration here hence:

> I personally lean towards the first option (simpler).

Agreed.

**Comment by FroggyFlox** [comment]
Thanks for your feedback.

I already got a `sed`-like little function that would tell us if there was any change required so I think I'll go that route.

**Comment by FroggyFlox** [comment]
Post 5fb66a6:

Manually set the following files to now-legacy paths:
```
rockdev:/opt/rockstor # cat /etc/samba/smb.conf 
(...)
    root preexec = "/opt/rockstor/bin/mnt-share test_share01"
    root preexec close = yes
    comment = Samba-Export
    path = /mnt2/test_share01
    browseable = yes
    read only = no
    guest ok = no
####END: Rockstor SAMBA CONFIG####
```
```
rockdev:/opt/rockstor # cat /etc/cron.d/rockstortab 
SHELL=/bin/bash
PATH=/sbin:/bin:/usr/sbin:/usr/bin
MAILTO=root
# These entries are auto generated by Rockstor. Do not edit.
42 3 * * 6 root /opt/rockstor/.venv/bin/st-system-power 1 \*-*-*-*-*-*
42 3 * * 5 root /opt/rockstor/.venv/bin/st-snapshot 2 \*-*-*-*-*-*
42 3 * * 5 root /opt/rockstor/.venv/bin/st-pool-scrub 3 \*-*-*-*-*-*
```

Then run `poetry run initrock`. Its logs end with:
```
2023-05-22 18:51:15,598: Establishing poetry path to binaries in local files...
2023-05-22 18:51:15,598: The replication_crontab (/etc/cron.d/replicationtab) could not be found
2023-05-22 18:51:15,599: samba_config updated.
2023-05-22 18:51:15,599: rockstor_crontab updated.
2023-05-22 18:51:15,600: DONE establishing poetry path to binaries in local files.
```

The files were correctly adjusted:
```
rockdev:/opt/rockstor # grep "/opt/rockstor/" {/etc/samba/smb.conf,/etc/cron.d/rockstortab}
/etc/samba/smb.conf:    root preexec = "/opt/rockstor/.venv/bin/mnt-share test_share01"
/etc/cron.d/rockstortab:42 3 * * 6 root /opt/rockstor/.venv/bin/st-system-power 1 \*-*-*-*-*-*
/etc/cron.d/rockstortab:42 3 * * 5 root /opt/rockstor/.venv/bin/st-snapshot 2 \*-*-*-*-*-*
/etc/cron.d/rockstortab:42 3 * * 5 root /opt/rockstor/.venv/bin/st-pool-scrub 3 \*-*-*-*-*-*
```

**Comment by FroggyFlox** [comment]
Re-running `poetry run initrock` ends with:
```
2023-05-22 19:07:06,148: Establishing poetry path to binaries in local files...
2023-05-22 19:07:06,148: The replication_crontab (/etc/cron.d/replicationtab) could not be found
2023-05-22 19:07:06,148: samba_config already looks good.
2023-05-22 19:07:06,148: rockstor_crontab already looks good.
2023-05-22 19:07:06,148: DONE establishing poetry path to binaries in local files.
```

**Comment by FroggyFlox** [comment]
I'll try to take some time to prepare a proper PR with testing later on, then :-).

**Comment by FroggyFlox** [comment]
Good catch... I'll wait for that PR to be merged before doing that then. I'll see if I can take the time to test your PR instead in this case as that'll be a better use of my time.

**Comment by FroggyFlox** [comment]
@phillxnet, a quick and nitpicky detail:

Based on our discussion, I'll add a step to enforce a mask of `600` to the crontabs. I can't decide on whether to use a nested dict, or follow your recent example and use a namedtuple.

Dict of dicts:
```python
LOCAL_FILES_DICT = {
    "samba_config": {
        "path": "/etc/samba/smb.conf",
        "mask": None
    },
    "rockstor_crontab": {
        "path": "/etc/cron.d/rockstortab",
        "mask": stat.S_IRUSR | stat.S_IWUSR
    },
    "replication_crontab": {
        "path": "/etc/cron.d/replicationtab",
        "mask": stat.S_IRUSR | stat.S_IWUSR
    },
}
```

namedtuple:
```python
LocalFile = namedtuple("LocalFile", "path mask")
LOCAL_FILES = {
    "samba_config": LocalFile(
        path="/etc/samba/smb.conf",
        mask=None
    ),
    "rockstor_crontab": LocalFile(
        path="/etc/cron.d/rockstortab",
        mask=stat.S_IRUSR | stat.S_IWUSR
    ),
    "replication_crontab": LocalFile(
        path="/etc/cron.d/replicationtab",
        mask=stat.S_IRUSR | stat.S_IWUSR
    ),
}
```

Both end up with similar code in `establish_poetry_paths()`. Any reason to prefer one over the other? I'm not sure it makes a real difference here given they are both relatively small but wanted to have your preference on that one if possible.

**Comment by phillxnet** [comment]
@FroggyFlox Repeating our side-channel discussion on the following:

> I can't decide on whether to use a nested dict, or follow your recent example and use a namedtuple.

I prefer the dictionary of named tuple as it's nicer to read and cleaner I think. Pretty sure it also uses less memory and is quicker, plus named tuples are immutable which makes them a form of constant within Python - pretty much what we are wanting to represent here I think. I.e. dict of named tuple reads nicer and is immutable in value. Plus I don't like the double dict references. Obviously we are a doocracy here so what-ever you find fits the bill really; as existing and working always beast idealised and non existent :). But I've been moving to using named tuples where that seems to fit the bill as it were.

**Comment by FroggyFlox** [comment]
@phillxnet, I have now rebased that branch post your recent merge.

A quick question regarding permissions for the files we correct here.
I kept the same mechanism as before:
1. create a temp file and write to it
2. change permissions:
    - if this is a file that we created (our crontabs): enforce a 600 mask.
    - if this is a file created by a system package (smb config): copy whatever mask the file already has
3. copy this temp file to its final destination

Now, based on our recent conversation, we have established that the default mask in this situation will be 600... so should we really keep step 2 above? I lean towards "yes" as it's an additional guarantee to have proper permissions and given we run that step only once, the cost should not factor here. Plus it allows us to use whatever permissions the source file has when we want that to happen.

What do you think?

**Comment by phillxnet** [comment]
@FroggyFlox Agreed re keeping step 2 re maintaining capability to 'copy' existing permissions. Currently our smb.conf is also a 600 as-is but that may in-fact be overly restrictive:
```
rleap15-4:~ # ls -la /etc/samba/smb.conf
-rw------- 1 root root 350 Mar  7 17:19 /etc/samba/smb.conf
```
But we are not aware of a problem there so lets not fix it. But we are aware of the cron files being less than optimal. So we should take the chance to improve these while we are here doing our Poetry path edit. Partial scope creep of sorts maybe, but it seems like a nice time to do this to improve folks stable updates as they transition from our last stable to the one we are working on now.

**Comment by FroggyFlox** [comment]
Thanks for the feedback, @phillxnet.

Regarding:

> Currently our smb.conf is also a 600 as-is but that may in-fact be overly restrictive:

Yes... Systemd seems to make sure the admin knows about it in the logs but it also does make it clear that it is NOT a problem so we should be good. It is the package's default after all.
```
May 25 11:53:47 rockdev systemd[1]: Configuration file /etc/systemd/system/smb.service is marked world-inaccessible. This has no effect as configuration data is accessible via APIs without restrictions. Proceeding anyway.
```

I'll go ahead and prepare the PR as it, then.

**Comment by phillxnet** [comment]
Closing as:
Fixed by #2558
